src/datasets/table.py

```diff
@@ -0,0 +1,591 @@
+        elif isinstance(blocks_part[0], Table):
+            table = pa.concat_tables([t.table for t in blocks_part])
+            blocks = [[t] for t in blocks_part]
+            return ConcatenationTable(table, blocks)
+        else:
+            tables_to_concat = []
+            blocks = blocks_part
+            for i, tables in enumerate(blocks):
+                for j, table in enumerate(tables):
+                    if j == 0:
+                        combined_table = table.table
+                    else:
+                        for name, col in zip(table.column_names, table.columns):
+                            combined_table = combined_table.append_column(name, col)
+                if i > 0 and combined_table.schema != tables_to_concat[0].schema:
+                    # re-order the columns to make the schema match and concat the tables
+                    names = tables_to_concat[0].schema.names
+                    arrays = [combined_table[name] for name in names]
+                    combined_table = pa.Table.from_arrays(arrays, names=names)
+                tables_to_concat.append(combined_table)
+            table = pa.concat_tables(tables_to_concat)
+            return ConcatenationTable(table, blocks)
+
+    @staticmethod
+    def from_tables(tables: List[Union[pa.Table, Table]]) -> "ConcatenationTable":
+        blocks = []
+        for table in tables:
+            if isinstance(table, pa.Table):
+                blocks.append([InMemoryTable(table)])
+            elif isinstance(table, ConcatenationTable):
+                blocks.extend(copy.deepcopy(table.blocks))
+            else:
+                blocks.append([table])
+        return ConcatenationTable.from_blocks(blocks)
+
+    @property
+    def _slices(self):
+        start = 0
+        for tables in self.blocks:
+            yield (start, start + len(tables[0]))
+            start += len(tables[0])
+
+    @wraps_arrow_table_method(pa.Table.slice)
+    def slice(self, offset=0, length=None):
+        table = self.table.slice(offset, length=length)
+        length = length if length is not None else self.num_rows - offset
+        blocks = []
+        for tables in self.blocks:
+            n_rows = len(tables[0])
+            if length == 0:
+                break
+            elif n_rows <= offset:
+                offset = offset - n_rows
+            elif n_rows <= offset + length:
+                blocks.append([t.slice(offset) for t in tables])
+                length, offset = length + offset - n_rows, 0
+            else:
+                blocks.append([t.slice(offset, length) for t in tables])
+                length, offset = 0, 0
+        return ConcatenationTable(table, blocks)
+
+    @wraps_arrow_table_method(pa.Table.filter)
+    def filter(self, mask, *args, **kwargs):
+        table = self.table.filter(mask, *args, **kwargs)
+        blocks = []
+        for (start, end), tables in zip(self._slices, self.blocks):
```

Comment:
the naming makes this quite hard to read, why is it called `end`?

Reply:
`end` is a very bad name here. Let me change it